src/utils/portals-common/portals-common.js

```diff
@@ -21,71 +24,152 @@ const getY = point => point[1];
 export const isPointInGeoJSONPolygon = isPointInPolygonWn(getX, getY);
 
 
+export const utcDateOnly = d => {
+  // Return a new Date object which is suitable for comparing UTC date only
+  // (fixed zero time component). One might suppose that the comparisons in
+  // the filter below also would work with local time (Date.setHours()), but
+  // they don't; only UTC works.
+  if (!d) {
+    return d;
+  }
+  const r = new Date(d);
+  r.setUTCHours(0, 0, 0, 0);
+  return r;
+};
+
+
+export const isNilOrAscending = (a, b, c) => {
+  // Allowing for nils (undefined, null), check that a <= b <= c.
+  return isNil(b) || ((isNil(a) || a <= b) && (isNil(c) || b <= c));
+};
+
+
+export const historyDateMatch = (
+  history, startDate, endDate, strict = true
+)  => {
+  // Return true if history observation dates match start and end date.
+  // Compare using dates only.
+  const minObsDate = utcDateOnly(history.min_obs_time);
+  const maxObsDate = utcDateOnly(history.max_obs_time);
+  startDate = utcDateOnly(startDate);
+  endDate = utcDateOnly(endDate);
+
+  if (strict) {
+    // Allowing for nils, check that
+    // minObsDate <= startDate <= endDate <= maxObsDate.
+    return (
+      isNilOrAscending(startDate, endDate)
+      && isNilOrAscending(minObsDate, startDate, maxObsDate)
+      && isNilOrAscending(minObsDate, endDate, maxObsDate)
+    );
+  }
+
+  // This is a variant of the looser legacy comparison. Allows nil start, end.
+  return (
+    !isNil(minObsDate) && !isNil(maxObsDate)
+    && isNilOrAscending(startDate, maxObsDate)
+    && isNilOrAscending(minObsDate, endDate)
+  );
+
+  // This is the looser comparison used by the legacy PDP portal.
+  // return (
+  //   !isNil(minObsDate) && !isNil(maxObsDate)
+  //   && !isNil(startDate) && !isNil(endDate)
+  //   && maxObsDate > startDate && minObsDate < endDate
+  // );
+};
+
+
+export const stationDateMatch = (
+  station, startDate, endDate, strict = true
+) => {
+  // TODO: Coalesce adjacent histories from a station. NB: tricky.
+  //  If we don't do this, and we use strict date matching, then a station with
+  //  several histories fully covering an interval will not be selected, even
+  //  though it should be. The question of what "adjacent" means is a bit tricky
+  //  ... would depend in part on history.freq to distinguish too-large gaps,
```

Comment:
We don't care about gaps. The users of the data necessarily has to assume that there will be gaps... that's just the nature of weather data collection.

Reply:
That's interesting. First, let's eliminate what I'm calling "strict" date matching, which is the condition 
`min_obs_time < uiStartDate < uiEndDate < max_obs_time`, plus allowance for nils
meaning complete containment in the observation interval. This is currently not used. 

Legacy PDP matching is currently used, which is the looser condition 
`min_obs_time < uiEndDate && uiStartDate < max_obs_time`, plus allowance for nils
meaning overlap with the observation interval

Currently, _for both legacy PDP and this app_, date matching is done _separately for each history in a station_. If this isn't right, it can be adjusted in this app.

Reply:
What I meant about gaps is this: Consider a station with multiple, say 2, histories spanning dates `A` to `B` (hx1) and `C` to `D` (hx2). 

Suppose the user specifies start and end dates `s`, `e` such that `A < s < B < C < e < D`. The legacy matching rule will operate as follows:
- hx1: `A < e && s < B === true`: match
- hx2 `C < e && s < D === true`: match

That's as desired.

But when we have, say, `A < B < s < e < C < D` then:
- hx1: `A < e && s < B === false`: no match
- hx2 `C < e && s < D === false`: no match

When start or end date fall into the gap between the histories' intervals, the matching fails. Even one of them in the gap means that one of the two histories will not match.

Questions: 
1. Does the gap problem matter?
2. Legacy PDP's effective definition of "station" is "history", and effectively ignores `meta_station` records. SDP treats history records linked by a common station as related.

Reply:
One further note: Because they are drawn from ` station_obs_stats_mv`, which is updated directly from observations, `min_obs_time` and `max_obs_time` are _never_ null (oops: except if there are no observations), and therefore the checking for nulls in the matching rules can (maybe) be simplified accordingly.

Reply:
I've moved this discussion to a [separate issue](https://github.com/pacificclimate/station-data-portal/issues/85).